prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -450,14 +452,155 @@ ssize_t rxr_pkt_wait_handshake(struct rxr_ep *ep, fi_addr_t addr, struct rxr_pee
+ * immediately.
+ *
+ * If data is copied by read. The progress engine will call the
+ * event handler.
+ */
+static inline
+ssize_t rxr_pkt_post_read_to_copy_data(struct rxr_ep *rxr_ep,
+				       struct rxr_pkt_entry *pkt_entry,
+				       char *data, size_t data_size,
+				       struct rxr_rx_entry *rx_entry,
+				       size_t data_offset)
+{
+	ssize_t err;
+	int iov_idx;
+	size_t iov_offset;
+	void *dest_buf;
+
+	assert(pkt_entry->x_entry == rx_entry);
+
+	err = rxr_locate_iov_pos(rx_entry->iov, rx_entry->iov_count, data_offset,
+				 &iov_idx, &iov_offset);
+
+	if (err) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"data_offset %ld out of range\n",
+			data_offset);
+		return err;
+	}
+
+	assert(iov_idx >= 0 && iov_idx < rx_entry->iov_count);
+	assert(efa_ep_is_cuda_mr(rx_entry->desc[iov_idx]));
+	assert(iov_offset + data_size <= rx_entry->iov[iov_idx].iov_len);
+
+	dest_buf = (char *)rx_entry->iov[iov_idx].iov_base + iov_offset;
+
+	pkt_entry->state = RXR_PKT_ENTRY_COPY_BY_READ;
+	assert(pkt_entry->mr);
+	assert(rxr_ep->rdm_self_addr != FI_ADDR_NOTAVAIL);
+
+	err = fi_read(rxr_ep->rdm_ep,
+		      dest_buf, data_size, rx_entry->desc[iov_idx],
+		      rxr_ep->rdm_self_addr,
+		      (uint64_t)data, fi_mr_key(pkt_entry->mr),
+		      pkt_entry);
+
+	if (OFI_UNLIKELY(err)) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ, "Failed to post read to copy data\n");
+		return err;
+	}
+
+	return 0;
+}
+
+static inline
+void rxr_pkt_handle_data_copied(struct rxr_ep *ep,
+				struct rxr_pkt_entry *pkt_entry)
+{
+	int pkt_type;
+
+	pkt_type = rxr_get_base_hdr(pkt_entry->pkt)->type;
+
+	switch (pkt_type) {
+	case RXR_DATA_PKT:
+	case RXR_READRSP_PKT:
+		rxr_data_pkt_handle_data_copied(ep, pkt_entry);
+		break;
+	case RXR_EAGER_MSGRTM_PKT:
+	case RXR_EAGER_TAGRTM_PKT:
+		rxr_pkt_handle_eager_rtm_data_copied(ep, pkt_entry);
+		break;
+	case RXR_MEDIUM_MSGRTM_PKT:
+	case RXR_MEDIUM_TAGRTM_PKT:
+		rxr_pkt_handle_medium_rtm_data_copied(ep, pkt_entry);
+		break;
+	case RXR_LONG_MSGRTM_PKT:
+	case RXR_LONG_TAGRTM_PKT:
+		rxr_pkt_handle_long_rtm_data_copied(ep, pkt_entry);
+		break;
+	case RXR_EAGER_RTW_PKT:
+		rxr_pkt_handle_eager_rtw_data_copied(ep, pkt_entry);
+		break;
+	case RXR_LONG_RTW_PKT:
+		rxr_pkt_handle_long_rtw_data_copied(ep, pkt_entry);
+		break;
+	default:
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"invalid control pkt type %d\n",
+			rxr_get_base_hdr(pkt_entry->pkt)->type);
+		assert(0 && "invalid control pkt type");
+		rxr_cq_handle_cq_error(ep, -FI_EIO);
+	}
+}
+
+void rxr_pkt_proc_data(struct rxr_ep *ep,
+		       struct rxr_rx_entry *rx_entry,
+		       size_t data_offset,
+		       struct rxr_pkt_entry *pkt_entry,
+		       char *data, size_t data_size)
+{
+	ssize_t err;
+	struct efa_mr *desc;
+
+	pkt_entry->x_entry = rx_entry;
+
+	if (OFI_UNLIKELY(rx_entry->rxr_flags & RXR_RECV_CANCEL) ||
+	    data_offset >= rx_entry->cq_entry.len) {
+		rxr_pkt_handle_data_copied(ep, pkt_entry);
+		return;
+	}
+
+	if (efa_ep_is_cuda_mr(rx_entry->desc[0])) {
```

Comment:
Would this not fully avoid gdrcopy or whatever alternative copy implementation is available?

Reply:
gdrcopy is not used by EFA yet. We need a separate patch to enable gdrcopy for EFA. My plan is in that patch implement a function ```rxr_ep_use_local_read()``` to determine whether local read is used.